plugins/renepay/pay.c

```diff
@@ -0,0 +1,1798 @@
+		json_add_sha256(ret, "payment_hash", &p->payment_hash);
+		json_add_u32(ret, "created_at", complete_created_at);
+		json_add_num(ret, "parts", complete_parts);
+
+		/* This payment was already completed, we don't keep record of
+		 * it twice. */
+		renepay->payment = tal_free(renepay->payment);
+
+		return command_finished(cmd, ret);
+	} else if (pending) {
+		p->groupid = last_pending_group_id;
+		renepay->next_partid = last_pending_partid+1;
+
+		p->total_sent = pending_sent;
+		p->total_delivering = pending_msat;
+
+		plugin_log(pay_plugin->plugin,LOG_DBG,
+			   "There are pending sendpays to this invoice. "
+			   "groupid = %ld, "
+			   "delivering = %s, "
+			   "last_partid = %ld",
+			   last_pending_group_id,
+			   type_to_string(tmpctx,struct amount_msat,&p->total_delivering),
+			   last_pending_partid);
+
+		if( first_pending_group_id != last_pending_group_id)
+		{
+			/* At least two pending groups for the same invoice,
+			 * this is weird, we better stop. */
+			renepay->payment = tal_free(renepay->payment);
+			return renepay_fail(renepay, PAY_IN_PROGRESS,
+					    "Payment is pending by some other request.");
+		}
+		if(amount_msat_greater_eq(p->total_delivering,p->amount))
+		{
+			/* Pending payment already pays the full amount, we
+			 * better stop. */
+			renepay->payment = tal_free(renepay->payment);
+			return renepay_fail(renepay, PAY_IN_PROGRESS,
+					    "Payment is pending with full amount already commited");
+		}
+	}else
+	{
+		p->groupid = (last_group==INVALID_ID  ? 1 : (last_group+1)) ;
+		renepay->next_partid=1;
+	}
+
+
+	struct out_req *req;
+	/* Get local capacities... */
+	req = jsonrpc_request_start(cmd->plugin, cmd, "listpeerchannels",
+				    listpeerchannels_done,
+				    listpeerchannels_done, renepay);
+	return send_outreq(cmd->plugin, req);
+}
+
+static struct command_result *json_pay(struct command *cmd,
+				       const char *buf,
+				       const jsmntok_t *params)
+{
+	const char *invstr;
+	const char *label;
+	const char *description;
+	struct sha256 * local_offer_id;
+ 	u64 invexpiry;
+ 	struct amount_msat *msat, *invmsat;
+	struct amount_msat *maxfee;
+	u64 *riskfactor_millionths;
+	u32 *maxdelay;
+	u64 *base_fee_penalty;
+	u64 *prob_cost_factor;
+	u64 *min_prob_success_millionths;
+	u32 *retryfor;
+
+#if DEVELOPER
+	bool *use_shadow;
+#endif
+
+	if (!param(cmd, buf, params,
+		   p_req("invstring", param_string, &invstr),
+ 		   p_opt("amount_msat", param_msat, &msat),
+ 		   p_opt("maxfee", param_msat, &maxfee),
+
+		   // MCF parameters
+		   // TODO(eduardo): are these parameters read correctly?
+		   p_opt_def("base_fee_penalty", param_millionths, &base_fee_penalty,10),
+ 		   p_opt_def("prob_cost_factor", param_millionths, &prob_cost_factor,10),
+		   p_opt_def("min_prob_success", param_millionths,
+		   	&min_prob_success_millionths,100000),// default is 10%
+
+		   p_opt_def("riskfactor", param_millionths,&riskfactor_millionths,1),
```

Comment:
Drop this, it's not useful?

Reply:
haven't started reading the full code yet but if risk factor is still related to CLTV stuff I would also suggest to remove it. I don't see how it is useful in payment delivery. Also I think most people never understood what it does (though I really like the idea)

Reply:
Drop the `riskfactor` or the `min_prob_success`?

Reply:
Both! Give them two knobs: max fee, time limit. That's enough

Reply:
What about leaving the options in developer mode only?

Reply:
I'll set the default value of `min_prob_success` to 90%, that means `minflow` will:
- first seek a solution that fits the fee budget,
- then tweak `mu` to try to meet the prob. budget `min_prob_success`
- then if both are met, then try to get as low fees as it can.